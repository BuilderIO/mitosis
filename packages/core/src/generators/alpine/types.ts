import { BaseTranspilerOptions } from '@/types/transpiler';

export interface ToAlpineOptions extends BaseTranspilerOptions {
  /**
   * use @on and : instead of `x-on` and `x-bind`
   */
  useShorthandSyntax?: boolean;
  /**
   * If true, the javascript won't be extracted into a separate script block.
   */
  inlineState?: boolean;
}

export type AlpineMetadata = {};
