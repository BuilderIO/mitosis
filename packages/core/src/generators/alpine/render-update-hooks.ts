import { curry } from 'lodash';
import { BaseHook, MitosisComponent } from '../../types/mitosis-component';

const extractCode = (hook: BaseHook) => hook.code;
function renderRootUpdateHook(hooks: BaseHook[], output: string) {
  if (hooks.length === 0) {
    return output;
  }
  const str = `onUpdate() {
        ${hooks.map(extractCode).join('\n')}
    }`;

  return output.replace(/,?(\s*})$/, `,\n${str}$1`);
}

function getRootUpdateHooks(json: MitosisComponent) {
  return (json.hooks.onUpdate ?? []).filter((hook) => hook.deps == '');
}

export function hasRootUpdateHook(json: MitosisComponent): boolean {
  return getRootUpdateHooks(json).length > 0;
}

export const renderUpdateHooks = curry((json: MitosisComponent, output: string) => {
  return renderRootUpdateHook(getRootUpdateHooks(json), output);
});

function getWatchHooks(json: MitosisComponent) {
  return (json.hooks.onUpdate ?? []).filter((hook) => hook.deps?.match(/state|this/));
}

export const hasWatchHooks = (json: MitosisComponent): boolean => {
  return getWatchHooks(json).length > 0;
};

function renderWatchHook(hook: BaseHook): string {
  const deps = (hook.deps ?? '')
    ?.slice(1)
    .slice(0, -1)
    .split(', ')
    .filter((dep) => dep.match(/state|this/));

  return deps
    .map(
      (dep) =>
        `this.$watch('${dep.replace(/(state|this)\./, '')}', (value, oldValue) => { ${
          hook.code
        } });`,
    )
    .join('\n');
}

export const renderWatchHooks = (json: MitosisComponent): string => {
  return hasWatchHooks(json) ? getWatchHooks(json).map(renderWatchHook).join('\n') : '';
};
