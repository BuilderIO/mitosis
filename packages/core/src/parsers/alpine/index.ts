import { parseExpression } from '@babel/parser';
import { MitosisComponent } from '@builder.io/mitosis';
import { parse as parseHTML } from 'node-html-parser';
import { createSingleBinding } from '../../helpers/bindings';
import { createMitosisNode } from '../../helpers/create-mitosis-node';
import { MitosisNode } from '../../types/mitosis-node';
import { AlpineAST, AlpineDirective, AlpineNode } from './types';

export function parseAlpine(html: string): MitosisComponent {
  const root = parseHTML(html);
  const ast: AlpineAST = {
    type: 'AlpineAST',
    root: parseNode(root),
    data: null,
  };

  // Find x-data directive
  const dataDirective = findDataDirective(ast.root);
  if (dataDirective) {
    ast.data = {
      type: 'AlpineData',
      value: dataDirective.value,
      parsed: parseData(dataDirective.value),
    };
  }

  return convertToMitosis(ast);
}

function findDataDirective(node: AlpineNode): { value: string } | null {
  const dataDirective = node.directives.find((d) => d.name === 'x-data');
  if (dataDirective) {
    return { value: dataDirective.value };
  }

  for (const child of node.children) {
    const result = findDataDirective(child);
    if (result) {
      return result;
    }
  }

  return null;
}

function parseData(value: string): Record<string, any> {
  try {
    // First try parsing as JSON
    return JSON.parse(value);
  } catch (e) {
    try {
      // Then try parsing as JavaScript expression
      const ast = parseExpression(value, {
        sourceType: 'module',
        plugins: ['jsx', 'typescript'],
      });
      return {
        expression: ast,
        value,
      };
    } catch (e) {
      // If both fail, return the raw value
      return { value };
    }
  }
}

function parseNode(node: any): AlpineNode {
  if (node.nodeType === 3) {
    // Text node
    return {
      type: 'AlpineNode',
      tagName: '',
      attributes: {},
      directives: [],
      children: [],
      text: node.text,
    };
  }

  if (node.nodeType === 8) {
    // Comment node
    return {
      type: 'AlpineNode',
      tagName: '',
      attributes: {},
      directives: [],
      children: [],
      text: `<!--${node.text}-->`,
    };
  }

  const directives: any[] = [];
  const attributes: Record<string, string> = {};

  // Parse attributes and directives
  const attrs = node.attributes;
  if (attrs) {
    Object.keys(attrs).forEach((name) => {
      const value = attrs[name];
      if (name.startsWith('x-')) {
        const [directiveName, ...modifiers] = name.split('.');
        directives.push({
          type: 'AlpineDirective',
          name: directiveName.slice(2), // Remove 'x-' prefix
          value,
          modifiers,
        });
      } else {
        attributes[name] = value;
      }
    });
  }

  // Parse child nodes
  const children: AlpineNode[] = [];
  if (node.childNodes) {
    for (const child of node.childNodes) {
      const parsedChild = parseNode(child);
      if (parsedChild.tagName || parsedChild.text) {
        children.push(parsedChild);
      }
    }
  }

  return {
    type: 'AlpineNode',
    tagName: node.tagName?.toLowerCase() || '',
    attributes,
    directives,
    children,
    text: undefined,
  };
}

function transformDirective(directive: AlpineDirective) {
  switch (directive.name) {
    case 'text':
      return {
        _text: createSingleBinding({ code: directive.value }),
      };
    case 'bind':
      // Handle x-bind:attribute="value" or :attribute="value"
      const attribute = directive.modifiers[0];
      if (!attribute) {
        throw new Error('x-bind directive requires an attribute name');
      }
      return {
        [attribute]: createSingleBinding({ code: directive.value }),
      };
    case 'on':
      // Handle x-on:event="handler" or @event="handler"
      const event = directive.modifiers[0] || 'click';
      // Convert Alpine event modifiers to Mitosis event handlers
      const modifiers = directive.modifiers.slice(1);
      let handlerCode = directive.value;

      // Handle event modifiers
      if (modifiers.includes('prevent')) {
        handlerCode = `(e) => { e.preventDefault(); ${handlerCode} }`;
      }
      if (modifiers.includes('stop')) {
        handlerCode = `(e) => { e.stopPropagation(); ${handlerCode} }`;
      }
      if (modifiers.includes('once')) {
        handlerCode = `(function() { let called = false; return (e) => { if (!called) { called = true; ${handlerCode} } } })()`;
      }
      if (modifiers.includes('self')) {
        handlerCode = `(e) => { if (e.target === e.currentTarget) { ${handlerCode} } }`;
      }
      if (modifiers.includes('window')) {
        handlerCode = `(e) => { if (e.target === window) { ${handlerCode} } }`;
      }
      if (modifiers.includes('document')) {
        handlerCode = `(e) => { if (e.target === document) { ${handlerCode} } }`;
      }
      if (modifiers.includes('outside')) {
        handlerCode = `(e) => { if (!e.currentTarget.contains(e.target)) { ${handlerCode} } }`;
      }

      // Handle debounce and throttle
      const debounceMatch = modifiers.find((m) => m.startsWith('debounce'));
      if (debounceMatch) {
        const ms = debounceMatch.split('.')[1] || '250';
        handlerCode = `(function() { let timeout; return (e) => { clearTimeout(timeout); timeout = setTimeout(() => { ${handlerCode} }, ${ms}); } })()`;
      }

      const throttleMatch = modifiers.find((m) => m.startsWith('throttle'));
      if (throttleMatch) {
        const ms = throttleMatch.split('.')[1] || '250';
        handlerCode = `(function() { let lastCall = 0; return (e) => { const now = Date.now(); if (now - lastCall >= ${ms}) { lastCall = now; ${handlerCode} } } })()`;
      }

      return {
        [`on${event.charAt(0).toUpperCase() + event.slice(1)}`]: createSingleBinding({
          code: handlerCode,
        }),
      };
    case 'if':
      return {
        _if: createSingleBinding({ code: directive.value }),
      };
    case 'for':
      // Handle x-for="item in items" or x-for="(item, index) in items" syntax
      const forMatch =
        directive.value.match(/^\((.*?),\s*(.*?)\)\s+in\s+(.*)$/) ||
        directive.value.match(/^(.*?)\s+in\s+(.*)$/);
      if (!forMatch) {
        throw new Error('Invalid x-for syntax');
      }

      const [_, item, index, items] = forMatch;
      return {
        each: createSingleBinding({ code: items }),
        forName: item,
        indexName: index,
      };
    case 'show':
      return {
        _show: createSingleBinding({ code: directive.value }),
      };
    case 'model':
      // Handle x-model="variable" for two-way binding
      // Also handle x-model modifiers like .number, .trim, .lazy
      const modelModifiers = directive.modifiers;
      let modelValue = directive.value;
      let modelHandler = `(e) => { ${directive.value} = e.target.value }`;

      if (modelModifiers.includes('number')) {
        modelHandler = `(e) => { ${directive.value} = Number(e.target.value) }`;
      }
      if (modelModifiers.includes('trim')) {
        modelHandler = `(e) => { ${directive.value} = e.target.value.trim() }`;
      }
      if (modelModifiers.includes('lazy')) {
        modelHandler = `(e) => { ${directive.value} = e.target.value }`;
      }

      return {
        value: createSingleBinding({ code: modelValue }),
        onChange: createSingleBinding({ code: modelHandler }),
      };
    case 'html':
      return {
        innerHTML: createSingleBinding({ code: directive.value }),
      };
    case 'cloak':
      // x-cloak is handled by CSS, no need for binding
      return {};
    case 'init':
      // x-init is handled by component initialization
      return {};
    case 'transition':
      // Handle x-transition directives
      const transitionModifier = directive.modifiers[0];
      if (transitionModifier) {
        return {
          [`transition${transitionModifier.charAt(0).toUpperCase() + transitionModifier.slice(1)}`]:
            createSingleBinding({ code: directive.value }),
        };
      }
      return {};
    case 'teleport':
      // Handle x-teleport directive
      return {
        _teleport: createSingleBinding({ code: directive.value }),
      };
    case 'ref':
      // Handle x-ref directive
      return {
        ref: createSingleBinding({ code: directive.value }),
      };
    default:
      return {
        [directive.name]: createSingleBinding({ code: directive.value }),
      };
  }
}

function transformNode(node: AlpineNode | null | undefined): MitosisNode {
  // Handle null or undefined nodes
  if (!node) {
    return createMitosisNode({
      '@type': '@builder.io/mitosis/node',
      name: 'div',
      properties: {},
      bindings: {},
      children: [],
      meta: {},
      scope: {},
    });
  }

  // Handle text and comment nodes
  if (node.text) {
    return createMitosisNode({
      '@type': '@builder.io/mitosis/node',
      name: 'div',
      properties: {},
      bindings: {
        _text: createSingleBinding({ code: `"${node.text}"` }),
      },
      children: [],
      meta: {},
      scope: {},
    });
  }

  const mitosisNode = createMitosisNode({
    '@type': '@builder.io/mitosis/node',
    name: node.tagName || 'div',
    properties: node.attributes || {},
    bindings: {},
    children: [],
    meta: {},
    scope: {},
  });

  // Transform directives
  if (node.directives) {
    node.directives.forEach((directive) => {
      Object.assign(mitosisNode.bindings, transformDirective(directive));
    });
  }

  // Transform children
  if (node.children) {
    mitosisNode.children = node.children.map((child) => transformNode(child));
  }

  return mitosisNode;
}

function convertToMitosis(ast: AlpineAST): MitosisComponent {
  const component: MitosisComponent = {
    '@type': '@builder.io/mitosis/component',
    name: 'AlpineComponent',
    imports: [],
    state: ast.data?.parsed || {},
    children: [],
    meta: {},
    hooks: {
      onMount: [],
      onEvent: [],
    },
    subComponents: [],
    context: {
      get: {},
      set: {},
    },
    props: {},
    refs: {},
    inputs: [],
  };

  component.children = [transformNode(ast.root)];

  return component;
}
