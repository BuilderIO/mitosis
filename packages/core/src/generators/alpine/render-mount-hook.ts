import { MitosisComponent } from '@/types/mitosis-component';
import { curry } from 'lodash';
import { stringifySingleScopeOnMount } from '../helpers/on-mount';
import { hasWatchHooks, renderWatchHooks } from './render-update-hooks';

function shouldRenderMountHook(json: MitosisComponent): boolean {
  return json.hooks.onMount.length > 0 || hasWatchHooks(json);
}

export const renderMountHook = curry((json: MitosisComponent, objectString: string) => {
  return shouldRenderMountHook(json)
    ? objectString.replace(
        /(?:,)?(\s*)(}\s*)$/,
        `, init() {
      ${renderWatchHooks(json)}
      ${stringifySingleScopeOnMount(json)}
    }$1$2`,
      )
    : objectString;
});
