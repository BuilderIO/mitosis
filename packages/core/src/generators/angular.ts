import dedent from 'dedent';
import { format } from 'prettier/standalone';
import { collectCss } from '../helpers/collect-styles';
import { fastClone } from '../helpers/fast-clone';
import { getRefs } from '../helpers/get-refs';
import { getStateObjectStringFromComponent } from '../helpers/get-state-object-string';
import { mapRefs } from '../helpers/map-refs';
import { renderPreComponent } from '../helpers/render-imports';
import { stripStateAndPropsRefs } from '../helpers/strip-state-and-props-refs';
import { selfClosingTags } from '../parsers/jsx';
import { MitosisNode } from '../types/mitosis-node';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../modules/plugins';
import isChildren from '../helpers/is-children';
import { getProps } from '../helpers/get-props';
import { kebabCase } from 'lodash';
import { stripMetaProperties } from '../helpers/strip-meta-properties';
import { removeSurroundingBlock } from '../helpers/remove-surrounding-block';
import { BaseTranspilerOptions, Transpiler } from '../types/config';
import { indent } from '../helpers/indent';

export interface ToAngularOptions extends BaseTranspilerOptions {}

const mappers: {
  [key: string]: (json: MitosisNode, options: ToAngularOptions) => string;
} = {
  Fragment: (json, options) => {
    return `<div>${json.children
      .map((item) => blockToAngular(item, options))
      .join('\n')}</div>`;
  },
};

export const blockToAngular = (
  json: MitosisNode,
  options: ToAngularOptions = {},
): string => {
  if (mappers[json.name]) {
    return mappers[json.name](json, options);
  }

  if (isChildren(json)) {
    return `<ng-content></ng-content>`;
  }

  if (json.properties._text) {
    return json.properties._text;
  }

  if (json.bindings._text) {
    return `{{${stripStateAndPropsRefs(json.bindings._text as string)}}}`;
  }

  let str = '';

  if (json.name === 'For') {
    str += `<ng-container *ngFor="let ${
      json.properties._forName
    } of ${stripStateAndPropsRefs(json.bindings.each as string)}">`;
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
    str += `</ng-container>`;
  } else if (json.name === 'Show') {
    str += `<ng-container *ngIf="${stripStateAndPropsRefs(
      json.bindings.when as string,
    )}">`;
    str += json.children
      .map((item) => blockToAngular(item, options))
      .join('\n');
    str += `</ng-container>`;
  } else {
    str += `<${json.name} `;

    // TODO: spread support for angular
    // if (json.bindings._spread) {
    //   str += `v-bind="${stripStateAndPropsRefs(
    //     json.bindings._spread as string,
    //   )}"`;
    // }

    for (const key in json.properties) {
      if (key.startsWith('$')) {
        continue;
      }
      const value = json.properties[key];
      str += ` ${key}="${value}" `;
    }
    for (const key in json.bindings) {
      if (key === '_spread') {
        continue;
      }
      if (key.startsWith('$')) {
        continue;
      }
      const value = json.bindings[key] as string;
      // TODO: proper babel transform to replace. Util for this
      const useValue = stripStateAndPropsRefs(value);

      if (key.startsWith('on')) {
        let event = key.replace('on', '').toLowerCase();
        if (
          event === 'change' &&
          json.name === 'input' /* todo: other tags */
        ) {
          event = 'input';
        }
        // TODO: proper babel transform to replace. Util for this
        const finalValue = removeSurroundingBlock(
          useValue.replace(/event\./g, '$event.'),
        );
        str += ` (${event})="${finalValue}" `;
      } else if (key === 'ref') {
        str += ` #${useValue} `;
      } else {
        str += ` [${key}]="${useValue}" `;
      }
    }
    if (selfClosingTags.has(json.name)) {
      return str + ' />';
    }
    str += '>';
    if (json.children) {
      str += json.children
        .map((item) => blockToAngular(item, options))
        .join('\n');
    }

    str += `</${json.name}>`;
  }
  return str;
};

export const componentToAngular =
  (options: ToAngularOptions = {}): Transpiler =>
  ({ component }) => {
    // Make a copy we can safely mutate, similar to babel's toolchain
    let json = fastClone(component);
    if (options.plugins) {
      json = runPreJsonPlugins(json, options.plugins);
    }

    const props = getProps(component);

    const refs = Array.from(getRefs(json));
    mapRefs(json, (refName) => `this.${refName}.nativeElement`);

    if (options.plugins) {
      json = runPostJsonPlugins(json, options.plugins);
    }
    let css = collectCss(json);
    if (options.prettier !== false) {
      css = tryFormat(css, 'css');
    }

    let template = json.children.map((item) => blockToAngular(item)).join('\n');
    if (options.prettier !== false) {
      template = tryFormat(template, 'html');
    }

    stripMetaProperties(json);

    const dataString = getStateObjectStringFromComponent(json, {
      format: 'class',
      valueMapper: (code) =>
        stripStateAndPropsRefs(code, { replaceWith: 'this.' }),
    });

    let str = dedent`
    import { Component ${refs.length ? ', ViewChild, ElementRef' : ''}${
      props.size ? ', Input' : ''
    } } from '@angular/core';
    ${renderPreComponent(json)}

    @Component({
      selector: '${kebabCase(json.name || 'my-component')}',
      template: \`
        ${indent(template, 8).replace(/`/g, '\\`').replace(/\$\{/g, '\\${')}
      \`,
      ${
        css.length
          ? `styles: [
        \`${indent(css, 8)}\`
      ],`
          : ''
      }
    })
    export default class ${component.name} {
      ${Array.from(props)
        .map((item) => `@Input() ${item}: any`)
        .join('\n')}

      ${refs
        .map((refName) => `@ViewChild('${refName}') ${refName}: ElementRef`)
        .join('\n')}

      ${
        !component.hooks.onMount
          ? ''
          : `ngOnInit() {
              ${stripStateAndPropsRefs(component.hooks.onMount.code, {
                replaceWith: 'this.',
              })}
            }`
      }

      ${
        !component.hooks.onUpdate?.length
          ? ''
          : `ngAfterContentChecked() {
              ${component.hooks.onUpdate.map((hook) =>
                stripStateAndPropsRefs(hook.code, {
                  replaceWith: 'this.',
                }),
              )}
            }`
      }

      ${
        !component.hooks.onUnMount
          ? ''
          : `ngOnDestroy() {
              ${stripStateAndPropsRefs(component.hooks.onUnMount.code, {
                replaceWith: 'this.',
              })}
            }`
      }

      ${dataString}
    }
  `;

    if (options.plugins) {
      str = runPreCodePlugins(str, options.plugins);
    }
    if (options.prettier !== false) {
      str = tryFormat(str, 'typescript');
    }
    if (options.plugins) {
      str = runPostCodePlugins(str, options.plugins);
    }
    return str;
  };

const tryFormat = (str: string, parser: string) => {
  try {
    return format(str, {
      parser,
      plugins: [
        // To support running in browsers
        require('prettier/parser-typescript'),
        require('prettier/parser-postcss'),
        require('prettier/parser-html'),
        require('prettier/parser-babel'),
      ],
      htmlWhitespaceSensitivity: 'ignore',
    });
  } catch (err) {
    console.warn('Could not prettify', { string: str }, err);
  }
  return str;
};
