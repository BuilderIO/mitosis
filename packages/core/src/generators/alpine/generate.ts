import { format } from 'prettier/standalone';
import { collectCss } from '../../helpers/styles/collect-css';
import { fastClone } from '../../helpers/fast-clone';
import { stripStateAndPropsRefs } from '../../helpers/strip-state-and-props-refs';
import { selfClosingTags } from '../../parsers/jsx';
import { checkIsForNode, ForNode, MitosisNode } from '../../types/mitosis-node';
import {
  runPostCodePlugins,
  runPostJsonPlugins,
  runPreCodePlugins,
  runPreJsonPlugins,
} from '../../modules/plugins';
import { stripMetaProperties } from '../../helpers/strip-meta-properties';
import { getStateObjectStringFromComponent } from '../../helpers/get-state-object-string';
import { BaseTranspilerOptions, TranspilerGenerator } from '../../types/transpiler';
import { dashCase } from '../../helpers/dash-case';
import { removeSurroundingBlock } from '../../helpers/remove-surrounding-block';
import { camelCase, curry, flow, flowRight as compose } from 'lodash';
import { getRefs } from '../../helpers/get-refs';
import { MitosisComponent } from '../../types/mitosis-component';
import { hasRootUpdateHook, renderUpdateHooks } from './render-update-hooks';
import { renderMountHook } from './render-mount-hook';
import { babelTransformCode } from '../../helpers/babel-transform';
import { replaceIdentifiers } from '../../helpers/replace-identifiers';

export interface ToAlpineOptions extends BaseTranspilerOptions {
  /**
   * use @on and : instead of `x-on` and `x-bind`
   */
  useShorthandSyntax?: boolean;
  /**
   * If true, the javascript won't be extracted into a separate script block.
   */
  inlineState?: boolean;
}

export const checkIsComponentNode = (node: MitosisNode): boolean =>
  node.name === '@builder.io/mitosis/component';

/**
 * Test if the binding expression would be likely to generate
 * valid or invalid liquid. If we generate invalid liquid tags
 * Shopify will reject our PUT to update the template
 */
export const isValidAlpineBinding = (str = '') => {
  return true;
  /*
  const strictMatches = Boolean(
    // Test for our `context.shopify.liquid.*(expression), which
    // we regex out later to transform back into valid liquid expressions
    str.match(/(context|ctx)\s*(\.shopify\s*)?\.liquid\s*\./),
  );

  return (
    strictMatches ||
    // Test is the expression is simple and would map to Shopify bindings	    // Test for our `context.shopify.liquid.*(expression), which
    // e.g. `state.product.price` -> `{{product.price}}	    // we regex out later to transform back into valid liquid expressions
    Boolean(str.match(/^[a-z0-9_\.\s]+$/i))
  );
  */
};

const removeOnFromEventName = (str: string) => str.replace(/^on/, '');
const removeTrailingSemicolon = (str: string) => str.replace(/;$/, '');
const trim = (str: string) => str.trim();

const replaceInputRefs = curry((json: MitosisComponent, str: string) => {
  getRefs(json).forEach((value) => {
    str = str.replaceAll(value, `this.$refs.${value}`);
  });

  return str;
});
const replaceStateWithThis = (str: string) => str.replaceAll('state.', 'this.');
const getStateObjectString = (json: MitosisComponent) =>
  flow(
    getStateObjectStringFromComponent,
    trim,
    replaceInputRefs(json),
    renderMountHook(json),
    renderUpdateHooks(json),
    replaceStateWithThis,
  )(json);

const bindEventHandlerKey = compose(dashCase, removeOnFromEventName);
const bindEventHandlerValue = compose(
  (x: string) =>
    replaceIdentifiers({
      code: x,
      from: 'event',
      to: '$event',
    }),
  removeTrailingSemicolon,
  trim,
  removeSurroundingBlock,
  stripStateAndPropsRefs,
);

const bindEventHandler =
  ({ useShorthandSyntax }: ToAlpineOptions) =>
  (eventName: string, code: string) => {
    const bind = useShorthandSyntax ? '@' : 'x-on:';
    return ` ${bind}${bindEventHandlerKey(eventName)}="${bindEventHandlerValue(code).trim()}"`;
  };

const mappers: {
  [key: string]: (json: MitosisNode, options: ToAlpineOptions) => string;
} = {
  For: (json, options) =>
    !(
      checkIsForNode(json) &&
      isValidAlpineBinding(json.bindings.each?.code) &&
      isValidAlpineBinding(json.scope.forName)
    )
      ? ''
      : `<template x-for="${json.scope.forName} in ${stripStateAndPropsRefs(
          json.bindings.each?.code,
        )}">
    ${(json.children ?? []).map((item) => blockToAlpine(item, options)).join('\n')}
    </template>`,
  Fragment: (json, options) => blockToAlpine({ ...json, name: 'div' }, options),
  Show: (json, options) =>
    !isValidAlpineBinding(json.bindings.when?.code)
      ? ''
      : `<template x-if="${stripStateAndPropsRefs(json.bindings.when?.code)}">
    ${(json.children ?? []).map((item) => blockToAlpine(item, options)).join('\n')}
    </template>`,
};

// TODO: spread support
const blockToAlpine = (json: MitosisNode | ForNode, options: ToAlpineOptions = {}): string => {
  if (mappers[json.name]) {
    return mappers[json.name](json, options);
  }

  // TODO: Add support for `{props.children}` bindings

  if (json.properties._text) {
    return json.properties._text;
  }

  if (json.bindings._text?.code) {
    return isValidAlpineBinding(json.bindings._text.code)
      ? `<span x-html="${stripStateAndPropsRefs(json.bindings._text.code as string)}"></span>`
      : '';
  }

  let str = `<${json.name} `;

  /*
  // Copied from the liquid generator. Not sure what it does. 
  if (
    json.bindings._spread?.code === '_spread' &&
    isValidAlpineBinding(json.bindings._spread.code)
  ) {
    str += `
          <template x-for="_attr in ${json.bindings._spread.code}">
            {{ _attr[0] }}="{{ _attr[1] }}"
          </template>
        `;
  }
  */

  for (const key in json.properties) {
    const value = json.properties[key];
    str += ` ${key}="${value}" `;
  }

  for (const key in json.bindings) {
    if (key === '_spread' || key === 'css') {
      continue;
    }
    const { code: value, type: bindingType } = json.bindings[key]!;
    // TODO: proper babel transform to replace. Util for this
    const useValue = stripStateAndPropsRefs(value);

    if (key.startsWith('on')) {
      str += bindEventHandler(options)(key, value);
    } else if (key === 'ref') {
      str += ` x-ref="${useValue}"`;
    } else if (isValidAlpineBinding(useValue)) {
      const bind = options.useShorthandSyntax && bindingType !== 'spread' ? ':' : 'x-bind:';
      str += ` ${bind}${bindingType === 'spread' ? '' : key}="${useValue}" `.replace(':=', '=');
    }
  }
  return selfClosingTags.has(json.name)
    ? `${str} />`
    : `${str}>${(json.children ?? []).map((item) => blockToAlpine(item, options)).join('\n')}</${
        json.name
      }>`;
};

export const componentToAlpine: TranspilerGenerator<ToAlpineOptions> =
  (options = {}) =>
  ({ component }) => {
    let json = fastClone(component);
    if (options.plugins) {
      json = runPreJsonPlugins(json, options.plugins);
    }
    const css = collectCss(json);
    stripMetaProperties(json);
    if (options.plugins) {
      json = runPostJsonPlugins(json, options.plugins);
    }

    const componentName = camelCase(json.name) || 'MyComponent';

    const stateObjectString = getStateObjectString(json);
    // Set x-data on root element
    json.children[0].properties['x-data'] = options.inlineState
      ? stateObjectString
      : `${componentName}()`;

    if (hasRootUpdateHook(json)) {
      json.children[0].properties['x-effect'] = 'onUpdate';
    }

    let str = css.trim().length ? `<style>${css}</style>` : '';
    str += json.children.map((item) => blockToAlpine(item, options)).join('\n');

    if (!options.inlineState) {
      str += `<script>
          ${babelTransformCode(`document.addEventListener('alpine:init', () => {
              Alpine.data('${componentName}', () => (${stateObjectString}))
          })`)}
        </script>`;
    }

    if (options.plugins) {
      str = runPreCodePlugins(str, options.plugins);
    }
    if (options.prettier !== false) {
      try {
        str = format(str, {
          parser: 'html',
          htmlWhitespaceSensitivity: 'ignore',
          plugins: [
            // To support running in browsers
            require('prettier/parser-html'),
            require('prettier/parser-postcss'),
            require('prettier/parser-babel'),
          ],
        });
      } catch (err) {
        console.warn('Could not prettify', { string: str }, err);
      }
    }
    if (options.plugins) {
      str = runPostCodePlugins(str, options.plugins);
    }
    return str;
  };
