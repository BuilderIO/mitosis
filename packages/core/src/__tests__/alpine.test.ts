import { componentToAlpine } from '../generators/alpine';
import { runTestsForTarget } from './shared';

describe('Alpine.js', () => {
  runTestsForTarget({ options: {}, target: 'alpine', generator: componentToAlpine });
  // runTestsForTarget({
  //   options: {
  //     standalone: true,
  //   },
  //   target: 'alpine',
  //   generator: componentToAlpine,
  // });
});
