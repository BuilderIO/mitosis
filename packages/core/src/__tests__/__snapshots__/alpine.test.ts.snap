// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alpine.js Remove Internal mitosis package 1`] = `
"<div x-data=\\"myBasicComponent()\\">
  Hello
  <span x-html=\\"name\\"></span>
  ! I can run in React, Qwik, Vue, Solid, or Liquid!
</div>
<script>
  document.addEventListener(\\"alpine:init\\", () => {
    Alpine.data(\\"myBasicComponent\\", () => ({ name: \\"PatrickJS\\" }));
  });
</script>
"
`;
